src/scalarstats.jl

```diff
@@ -274,25 +274,25 @@ If `normalize` is set to `true`, the MAD is multiplied by
 `1 / quantile(Normal(), 3/4) ≈ 1.4826`, in order to obtain a consistent estimator
 of the standard deviation under the assumption that the data is normally distributed.
 """
-function mad!(v::AbstractArray{T};
+function mad!(v::AbstractArray{<:Real};
               center::Real=median!(v),
               normalize::Union{Bool,Nothing}=true,
-              constant=nothing) where T<:Real
-    for i in 1:length(v)
-        @inbounds v[i] = abs(v[i]-center)
-    end
-    k = 1 / (-sqrt(2 * one(T)) * erfcinv(3 * one(T) / 2))
-    if normalize === nothing
+              constant=nothing)
+    isempty(v) && throw(ArgumentError("mad is not defined for empty arrays"))
+    v .= abs.(v .- center)
+    m = median!(v)
+    if normalize isa Nothing
         Base.depwarn("the `normalize` keyword argument will be false by default in future releases: set it explicitly to silence this deprecation", :mad)
         normalize = true
     end
-    if constant !== nothing
+    if !isa(constant, Nothing)
         Base.depwarn("keyword argument `constant` is deprecated, use `normalize` instead or apply the multiplication directly", :mad)
-        constant * median!(v)
+        m * constant
     elseif normalize
-        k * median!(v)
+        ## inv(sqrt(2) * erfinv(.5)) From https://www.wolframalpha.com/input/?i=1+%2F+(sqrt(2)+*+erfinv(.5))
+        m * oftype(m, BigFloat("1.482602218505601860547076529360423431326703202590312896536266275245674447622701"))
```

Comment:
Actually for performance we should probably define a custom `Irrational` with special `convert` methods for `Float64` and `Float32`. Or maybe just define a function taking `m` and returning the ajusted value, with a general `BigFloat` method and two optimized methods for `Float64` and `Float32`.

Reply:
No, I meant we should define an `Irrational` for the constant `k`.

Reply:
I am *not* speaking about `Vector{<:Irrational}`. I'm just saying that `oftype(m, BigFloat("1.48260..."))` isn't very efficient, and that using a custom `Irrational` with specialized `convert` methods could be a better approach. That's how this kind of situation is handled in Base.

Reply:
Got it. I couldn't find in Base how exactly to make an irrational number...
I did a few attempts at `@irrational(sym, val, def)`, but could use some guidance there.

Reply:
I guess you'd need to define methods for an `Irrational` with a particular symbol as type parameter. But given that it's only used internally it's probably simpler not to use `Irrational` and just define a function taking a type and returning the value converted to that type, with specialized methods.

Reply:
The `Irrational` way actually works quite well. I learned how to do implement it from the StatsFuns code. If it compiles and all we can go this way and see if we can use it for the depreciated code as well.

Reply:
Cool! I don't think we should use it for the deprecation, since people should just omit the argument if they want to use the constant.

It would be clearer to use a longer name than `k` though.

Reply:
How about `MAD_Normal_Scale_Factor`?

Reply:
Usually Julia objects use lowercase names; also no need to be too long either. So maybe `mad_constant`?

Reply:
The value would be a `DataType` which usually has upper case letters except that for `T<:Irrational` it seems the convention is to use lowercase as values. Sounds good.